portfolio/src/main/webapp/script.js

```diff
@@ -28,103 +28,214 @@ function addRandomGreeting() {
+/**
+ * Clears the current highlighted text nodes
+ * Will optimize to actually remove the span rather than the class
+ */
+function clear(){
+    let highlightedList = document.querySelectorAll('.highlight');
 
-    return source.substring(startIndex, endIndex)
+    highlightedList.forEach(function(el) {
+        el.replaceWith(el.innerText);
+    });
+    
 }
 
 /**
-  * Gets HTML id from given word
-  */
-function retrieveId(string, source){
-    var searchForId = string.search("id=\"");
-    if (!searchForId){
+ * Handles clearing old nodes and end of search run,
+ * as well as calls the highlighting and scrolling fns
+ */
+function highlightHandler(matches, input, index){
+    if (index === matches.length && index != 0){
+        clear();
+        setHidden('nextbutton', true);
+        match_index = 0; // resetting the index
+        document.body.scrollIntoView({behavior: "smooth"}); // back to the top
         return;
     }
-    const startIndex = searchForId+4;
-    var retString = string.charAt(startIndex);
 
-    for(var i = startIndex+1; string.charAt(i) != "\""; i++){
-        retString += string.charAt(i);
+    if (index > 0){
+        clear();
     }
+    
+    console.log(index);
+    console.log(matches);
+    console.log(matches[index]);
 
-    return retString;
+    // highlight and scroll to every element
+    highlighter(matches[index], input);
+    
+    console.log(matches[index]);
+
+    let els = document.getElementsByClassName('highlight');
+    els[0].scrollIntoView({behavior: "smooth"}); 
+
+    match_index++; // go onto the next value
 }
 
+
 /**
-  * Runs the search by getting the source HTML, getting the appropriate tag, and scrolling towards that id
-  */
-function runSearch(input){
-    var source = retrieveHtml();
+ * Highlights a given text node
+ */
+function highlighter(node, input){
+    const nodeText = node.textContent;
+
+    let foundIndices = findNeedleInHaystack(input, nodeText);
+
+    console.log(foundIndices);
+    if (foundIndices == [] || foundIndices == null){
+        throw "error with found indices";
+    }
+
+    let stringBefore = nodeText.substring(0, foundIndices[0]);
+    let stringAfter = nodeText.substring(input.length+foundIndices[0]);
+    let highlightedString = nodeText.substring(foundIndices[0], input.length+foundIndices[0]);
+    
+    let fragments = nodeText.split(highlightedString);
+
+    const newNode = document.createElement('span');
+    newNode.setAttribute('class', 'highlight');
+    newNode.textContent = highlightedString;
 
-    var foundIndex = source.indexOf(input);
-    if (foundIndex == -1){return};
+    for (let i = 1; i < fragments.length; i+=2){
+        fragments.splice(i, 0, newNode);
+    }
+
+    //node.replaceWith(...fragments);
+    node.replaceWith(stringBefore, newNode, stringAfter);
+}
+
+/**
+ * Sets the stage for the actual search to take place,
+ * handles gathering data and checks criteria are met
+ */
+function searchSetup(input){ 
```

Comment:
name this parameter `retrievedInput`. Also, I think just naming this function `search` would be more appropriate. `searchSetup` implies it should only ever be called once, and that it would set up the event handlers on the search box.